linera-client/src/benchmark.rs

```diff
@@ -162,6 +217,68 @@ where
             });
         }
 
+        let metrics_watcher = if let Some(health_check_endpoints) = health_check_endpoints {
+            let metrics_addresses = health_check_endpoints
+                .split(',')
+                .map(|address| address.trim())
+                .map(|address| {
+                    let parts = address.split(':').collect::<Vec<_>>();
+                    ensure!(
+                        parts.len() == 2,
+                        BenchmarkError::ParseValidatorMetricsAddress(address.to_string())
+                    );
+                    let host = parts[0].to_owned();
+                    let port = parts[1]
+                        .parse::<u16>()
+                        .map_err(BenchmarkError::ParseValidatorMetricsPort)?;
+                    Ok(format!("http://{}:{}/metrics", host, port))
+                })
+                .collect::<Result<Vec<_>, BenchmarkError>>()?;
+
+            let mut previous_histogram_snapshots: HashMap<String, HistogramSnapshot> =
+                HashMap::new();
+            let scrapes = Self::get_scrapes(&metrics_addresses).await?;
+            for (metrics_address, scrape) in scrapes {
+                previous_histogram_snapshots.insert(
+                    metrics_address,
+                    Self::parse_histogram(&scrape, LATENCY_METRIC_PREFIX)?,
+                );
+            }
+
+            let watcher_shutdown_notifier = shutdown_notifier.clone();
+            let metrics_watcher: task::JoinHandle<Result<(), BenchmarkError>> = tokio::spawn(
+                async move {
+                    let mut health_interval = time::interval(time::Duration::from_secs(5));
+                    let mut shutdown_interval = time::interval(time::Duration::from_secs(1));
+                    loop {
+                        tokio::select! {
+                            biased;
+                            _ = health_interval.tick() => {
+                                let result = Self::validators_healthy(&metrics_addresses, &mut previous_histogram_snapshots).await;
+                                if result.is_err() || !result? {
+                                    info!("Shutting down benchmark...");
+                                    watcher_shutdown_notifier.cancel();
+                                    break;
+                                }
+                            }
+                            _ = shutdown_interval.tick() => {
+                                if watcher_shutdown_notifier.is_cancelled() {
+                                    info!("Shutdown signal received, stopping metrics watcher");
+                                    break;
+                                }
+                            }
+                        }
+                    }
+
+                    Ok(())
+                },
+            );
+
+            Some(metrics_watcher)
```

Comment:
Maybe the whole `if Some` branch should be extracted into a function. (`create_metrics_watcher`?)